### PR TITLE
hng64 : add priority z sort mode for sprites

### PR DESCRIPTION
**Pull request by Happy-yappH**

### CODE REVIEW

File: src/mame/video/hng64_sprite.hxx [hunk 1]
```diff
@@ -56,19 +58,63 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
...
+			source += 8;
+		}
+		else
+		{
+			if (m_spriteregs[0] & 0x01000000)
+		// This flips between ingame and other screens for roadedge, where the sprites which are filtered definitely needs to change and the game explicitly swaps the values in the sprite list at the same time.
+		// m_spriteregs[2] could also play a part as it also flips between 0x00000000 and 0x000fffff at the same time
+		// Samsho games also set the upper 3 bits which could be related, samsho games still have some unwanted sprites (but also use the other 'sprite clear' mechanism)
+		// Could also be draw order related, check if it inverts the z value?
+			{
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:74]: Please don’t put anything between the control statement and the start of the controlled compound statement – it’s confusing, especially when the comment isn’t indented to match the code like this.

File: src/mame/video/hng64_sprite.hxx [hunk 2]
```diff
@@ -56,19 +58,63 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
...
+				// sort by spriteram entry order
+				// not sure if z priority field has another interpretation here
+				if ((source[2]&0x07ff0000)>>16 != 0x7ff)
+				{
+					sortlist.insert(std::pair <int, uint32_t*> (finish - source, source));
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:79]: Use `emplace` for stuff like this – it’s a lot less clutter.

File: src/mame/video/hng64_sprite.hxx [hunk 3]
```diff
@@ -46,6 +46,8 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
 	uint32_t *source = m_spriteram;
 	uint32_t *finish = m_spriteram + 0xc000/4;
 
+	std::multimap<int, uint32_t*> sortlist;
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:49]: Is a multimap the best structure for this?  It’s going to cause one allocation per element.  You might be better off using a vector, pre-allocating, and using either `lower_bound` or `upper_bound` to find the right place to insert things.

Comment (Happy-yappH) [reply src/mame/video/hng64_sprite.hxx:49]: The list does have a predetermined maximum size, but needs to allow multiple elements to have the same key.

Comment (cuavas) [reply src/mame/video/hng64_sprite.hxx:49]: Yeah, but with a vector you only need to do one allocation, you can choose whether you want to insert elements with the same key at the first or last valid location for that key (by using `lower_bound` or `upper_bound` to choose the location), you still have the same O(log2(n)) performance to find the insert location, you have better locality, and the O(n) “slide” when inserting at a location other than the end is pretty cheap when the elements are trivially copyable.

File: src/mame/video/hng64_sprite.hxx [hunk 4]
```diff
@@ -56,19 +58,63 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
...
+				}
+				else
+					source += 8;
+			}
+			else	
+			{
+				// sort by z priority
+				if ((source[2]&0x07ff0000)>>16 != 0)
+				{
+					sortlist.emplace(std::pair <int, uint32_t*> ((source[2]&0x07ff0000)>>16, source));
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:93]: You understand `emplace`, right?  It forwards its arguments to the element constructor.

File: src/mame/video/hng64_sprite.hxx [hunk 5]
```diff
@@ -50,25 +50,57 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
...
+				else
+					source += 8;
+			}
+			else
+				source += 8;
+		}
+	}
+
+	if (zsort)
+		std::sort(m_spritelist.begin(), m_spritelist.end(), std::greater<>());
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:90]: This is sorting the pair lexically, i.e. by the key first and then by the pointer.  Is that what you actually want?  I would’ve expected you’d want to use `stable_sort` and only compare the key.

File: src/mame/video/hng64_sprite.hxx [hunk 6]
```diff
@@ -50,25 +50,57 @@ void hng64_state::draw_sprites(screen_device &screen, bitmap_rgb32 &bitmap, cons
...
+
+	for(auto it = m_spritelist.begin(); it != m_spritelist.end(); it++)
```
Comment (cuavas) [inline src/mame/video/hng64_sprite.hxx:92]: Why not use range-based `for` loop syntax here?

File: src/mame/video/hng64_sprite.hxx (hunk 5 above)
Comment (Happy-yappH) [reply src/mame/video/hng64_sprite.hxx:90]: At the moment, I'm not sure if their is any precedence for sprites with the same priority key, so it probably wouldn't matter.  If sorting by just the key would be faster, that would be the preferable option.

File: src/mame/video/hng64_sprite.hxx (hunk 6 above)
Comment (Happy-yappH) [reply src/mame/video/hng64_sprite.hxx:92]: Because I am unaware of such syntax.  I will take a look.

File: src/mame/video/hng64_sprite.hxx (hunk 5 above)
Comment (cuavas) [reply src/mame/video/hng64_sprite.hxx]: Sorting by key only is faster – you can do it by using a lambda for the comparator, something like `[] (auto const &a, auto const &b) { return a.first < b.first; }` would work.

**Pull request merged**